cpp/modmesh/io/gmsh.hpp

```diff
@@ -0,0 +1,397 @@
+        // clang-format on
+    }
+#undef MM_DECL_SWITCH_ELEMENT_TYPE
+#undef VEC
+}
+
+class Gmsh
+    : public NumberBase<int32_t, double>
+{
+    using number_base = NumberBase<int32_t, double>;
+    using int_type = typename number_base::int_type;
+    using uint_type = typename number_base::uint_type;
+    using real_type = typename number_base::real_type;
+
+public:
+    explicit Gmsh(const std::string & file_path)
+    {
+        bool meta_enter = false, node_enter = false, element_enter = false;
+        // clang-format off
+        std::unordered_map<std::string, std::function<void()>> keyword_handler = {
+            {"$MeshFormat", [this, &meta_enter]() { _load_meta(); meta_enter = true; }},
+            {"$Nodes", [this, &node_enter]() { _load_nodes(); node_enter = true; }},
+            {"$Elements", [this, &element_enter]() { _load_elements(); element_enter = true; }},
+            {"$PhysicalNames", [this]() { _load_physical(); }}
+        };
+
+        stream.open(file_path, std::ios::in);
+        if (!stream.good())
+        {
+            throw std::invalid_argument(Formatter() << file_path << " path invalid.");
+        }
+
+        std::string line;
+        while (std::getline(stream, line))
+        {
+            // Using a finite state machine to check the input msh file format is valid or not.
+            // $ is a keyword to trigger state transition.
+            if (line.find('$') != std::string::npos)
+            {
+                auto it = keyword_handler.find(line);
+                if (it != keyword_handler.end())
+                {
+                    if (_isValidTransition(it->first))
+                    {
+                        it->second();
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+                else
+                {
+                    break;
+                }
+            }
+        }
+
+        // MeshFormat, Nodes and Elements section is mandatory in gmsh msh file,
+        // therefore need to check these sections are exist or not otherwise
+        // modmesh may crash during mesh processing.
+        if (!(meta_enter && node_enter && element_enter))
+        {
+            throw std::invalid_argument("Incorrect msh file format.");
+        }
+    }
+
+    ~Gmsh()
+    {
+        stream.close();
+    }
+
+    Gmsh() = delete;
+    Gmsh(Gmsh const & other) = delete;
+    Gmsh(Gmsh && other) = delete;
+    Gmsh & operator=(Gmsh const & other) = delete;
+    Gmsh & operator=(Gmsh && other) = delete;
+
+    std::shared_ptr<StaticMesh> toblock(void);
+
+private:
+    enum class FormatState
+    {
+        BEGIN,
+        META_END,
+        PYHSICAL_NAME_END,
+        NODE_END,
+        ELEMENT_END
+    };
+
+    // Check the finite state machine state transition is valid or not to check msh file format is correct
+    bool _isValidTransition(const std::string s)
```

Comment:
Do not prepend `_` to any C++ symbols.  You already made them private and mixing the C++ access with Python private convention is a bad idea.

Reply:
I have removed `_` from these private member function.